searvey/coops.py

```diff
@@ -994,13 +1050,391 @@ def get_coops_stations(
+    return formatted
+
+
+def _generate_urls(
+    station_id: str,
+    start_date: pd.Timestamp,
+    end_date: pd.Timestamp,
+    product: COOPS_Product = COOPS_Product.WATER_LEVEL,
+    datum: COOPS_TidalDatum = COOPS_TidalDatum.MSL,
+    units: COOPS_Units = COOPS_Units.METRIC,
+    interval: COOPS_Interval = COOPS_Interval.H,
+    bin_no: int = 0,
+) -> list[str]:
+    if end_date < start_date:
+        raise ValueError(f"'end_date' must be after 'start_date': {end_date} vs {start_date}")
+    if end_date == start_date:
+        return []
+    duration = end_date - start_date
+    periods = duration.days // MAX_COOPS_ITERVAL + 2
+    urls = []
+    date_range = pd.date_range(start_date, end_date, periods=periods, unit="us", inclusive="both")
+    params = {
+        "station": station_id,
+        "product": product.value,
+        "datum": datum.value,
+        "units": units.value,
+        "time_zone": "gmt",  # We always work with UTC/GMT
+        "interval": interval.value,
+        "format": "json",
+        "application": "oceanmodeling/stormevents",
+    }
+    if product in [COOPS_Product.CURRENTS, COOPS_Product.CURRENTS_PREDICTIONS]:
+        params["bin"] = bin_no
+    for start, stop in itertools.pairwise(date_range):
+        params["begin_date"] = _coops_date(start)
+        params["end_date"] = _coops_date(stop)
+        req = requests.Request("GET", COOPS_BASE_URL, params=params)
+        urls.append(str(req.prepare().url))
+    return urls
+
+
+def _normalize_df(df: pd.DataFrame, product: COOPS_Product) -> pd.DataFrame:
+    # TODO: Add more info (datum, unit, tz)?
+
+    nos_id = df.attrs["station_id"]
+    normalized = df.rename(columns=COOPS_ProductFieldsNameMap[product])
+    logger.debug("%s: df contains the following columns: %s", nos_id, normalized.columns)
+
+    normalized[normalized == ""] = numpy.nan
+    normalized = normalized.astype(
+        {k: v for k, v in COOPS_ProductFieldTypes.items() if k in df.columns}, errors="ignore"
+    )
+    normalized["time"] = pandas.to_datetime(normalized["time"])
+    normalized.set_index("time", inplace=True)
+
+    return normalized
+
+
+def _parse_json(content: str, station_id: str, product: COOPS_Product) -> pd.DataFrame:
+    content_json = json.loads(content)
+    data = []
+    if product == COOPS_Product.CURRENTS_PREDICTIONS:
+        data = content_json["current_predictions"]["cp"]
+    else:
+        data = content_json["data"]
+
+    df = pd.DataFrame(data)
+    df.attrs["station_id"] = f"COOPS-{station_id}"
+    df = _normalize_df(df, product)
+    return df
+
+
+def _group_results(
+    station_ids: abc.Collection[str],
+    parsed_responses: list[multifutures.FutureResult],
+) -> dict[str, pd.DataFrame]:
+    # Group per COOPS code
+    df_groups = collections.defaultdict(list)
+    for item in parsed_responses:
+        df_groups[item.kwargs["station_id"]].append(item.result)  # type: ignore[index]
+
+    # Concatenate dataframes and remove duplicates
```

Comment:
Are there duplicates in COOPS?

Reply:
I created this first commit by copying all IOC updates and modifying parts that were required. I didn't spend a lot of time thinking about the grouping. I think it'd be a good idea to discuss whether some of these are needed or not.

Reply:
@pmav99 I'm trying to decide what parts to keep here. What is the role of grouping for IOC? What is it grouping?

Reply:
Nevermind, I just realized it's combining periods that we divided by max data length. I updated the code since there's no duplicate in COOPS as you said.